src/control/server/harness.go

```diff
@@ -231,6 +233,124 @@ func (h *IOServerHarness) startInstances(ctx context.Context, membership *system
 	return nil
 }
 
+// StopInstances will signal harness-managed instances.
+//
+// Iterate over instances and call Stop(sig) on each, return when all instances
+// exit or err context is done. Error map returned for stop attempt failures so
+// can continue to return results and update state.
+func (h *IOServerHarness) stopInstances(ctx context.Context, signal os.Signal, rankList ...*ioserver.Rank) (map[*ioserver.Rank]error, error) {
+	if !h.IsStarted() {
+		return nil, nil
+	}
+	if signal == nil {
+		return nil, errors.New("nil signal")
+	}
+
+	instances := h.Instances()
+	type rankRes struct {
+		rank *ioserver.Rank
+		ps   *os.ProcessState
+		err  error
+	}
+	resChan := make(chan rankRes, len(instances))
+	stopping := 0
+	for _, instance := range instances {
+		rank, err := instance.GetRank()
+		if err != nil {
+			return nil, err
+		}
+
+		if !checkRankList(*rank, rankList) {
+			h.log.Debugf("rank %s not in requested list, skipping...", rank)
+			continue // filtered out, no result expected
+		}
+
+		if !instance.IsStarted() {
+			continue
+		}
+
+		go func(i *IOServerInstance) {
+			h.log.Debugf("%s rank %s", signal, rank)
+
+			ps, err := i.Stop(signal) // blocks until process.Wait() returns
+			resChan <- rankRes{rank: rank, ps: ps, err: err}
+		}(instance)
+		stopping++
+	}
+
+	stopErrors := make(map[*ioserver.Rank]error)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case result := <-resChan:
+			stopping--
+			if result.err != nil {
+				stopErrors[result.rank] = result.err
+			} else {
+				h.log.Debugf("process state on exit: %s", result.ps)
+			}
+			if stopping == 0 {
+				return stopErrors, nil
+			}
+		}
+	}
+}
+
+func (h *IOServerHarness) getInstanceStartedResults(rankList []*ioserver.Rank, desiredState system.MemberState, action string, stopErrs map[*ioserver.Rank]error) (system.MemberResults, error) {
+	results := make(system.MemberResults, 0, maxIOServers)
+	for _, i := range h.Instances() {
+		rank, err := i.GetRank()
+		if err != nil {
+			return nil, err
+		}
+
+		if !checkRankList(*rank, rankList) {
+			continue // filtered out, no result expected
+		}
+
+		state := system.MemberStateStarted
+		if !i.IsStarted() {
+			state = system.MemberStateStopped
+		}
+
+		var extraErrMsg string
+		if stopErrs != nil {
```

Comment:
it doesn't really matter if it's more than zero , as none of the ranks will match an empty list. we do need to check if the input map is nil because we don't want to look up a nil map

Reply:
Sure, but my point is that by checking the length of the map you can a) avoid accessing a nil map and b) avoid iterating if the map isn't nil but is empty. Using len() on the zero value of iterable types yields 0: https://play.golang.org/p/IdkXM7_WhWf

Reply:
okay understood